File: csharp/Solitaire/Klondike/Move/RecycleStockPileMove.cs
using Silnith.Game.Deck;
using System;
using System.Collections.Generic;

namespace Silnith.Game.Klondike.Move
{
    /// <summary>
    /// A move that recycles the stock pile.  This sets the current index into the
    /// stock pile back to zero.
    /// </summary>
    public class RecycleStockPileMove : ISolitaireMove, IEquatable<RecycleStockPileMove?>
    {
        /// <summary>
        /// Finds all recycle stock pile moves for a given board.
        /// </summary>
        /// <remarks>
        /// <para>
        /// This will either contain one move or zero.
        /// </para>
        /// </remarks>
        /// <param name="board">The board to examine.</param>
        /// <returns>An enumerable of moves.</returns>
        public static IEnumerable<ISolitaireMove> FindAllMovesForBoard(Board board)
        {
            if (board.CanRecycleStockPile())
            {
                return new List<ISolitaireMove>(1)
                {
                    new RecycleStockPileMove(board),
                };
            }
            else
            {
                return Array.Empty<ISolitaireMove>();
            }
        }

        /// <summary>
        /// The index into the stock pile before the move is applied.
        /// </summary>
        /// <remarks>
        /// <para>
        /// For a legal move, this will be equal to the size of the stock pile.
        /// </para>
        /// </remarks>
        public int SourceIndex
        {
            get;
        }

        /// <inheritdoc/>
        public bool HasCards
        {
            get
            {
                return false;
            }
        }

        /// <inheritdoc/>
        public IReadOnlyList<Card> Cards
        {
            get
            {
                throw new NotImplementedException();
            }
        }

        /// <summary>
        /// Constructs a new move that recycles the stock pile.
        /// </summary>
        /// <param name="sourceIndex">The index into the stock pile before the move.</param>
        public RecycleStockPileMove(int sourceIndex)
        {
            SourceIndex = sourceIndex;
        }

        /// <summary>
        /// Constructs a new move that recycles the stock pile.
        /// </summary>
        /// <param name="board">The board from which to get the stock pile index.</param>
        public RecycleStockPileMove(Board board) : this(board.StockPileIndex)
        {
        }

        /// <inheritdoc/>
        public Board Apply(Board board)
        {
            return new Board(board.Columns, board.StockPile, 0, board.Foundation);
        }

        /// <inheritdoc/>
        public override bool Equals(object? obj)
        {
            return Equals(obj as RecycleStockPileMove);
        }

        /// <inheritdoc/>
        public bool Equals(RecycleStockPileMove? other)
        {
            return other != null &&
                   SourceIndex == other.SourceIndex;
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            return HashCode.Combine(SourceIndex);
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            return $"Recycle stock pile from index {SourceIndex} to the beginning.";
        }
    }
}

File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/RecycleStockPileMove.java
package org.silnith.game.solitaire.move;

import java.util.List;

import org.silnith.deck.Card;
import org.silnith.game.solitaire.Board;


/**
 * A move that recycles the stock pile.  This sets the current index into the
 * stock pile back to zero.
 */
public class RecycleStockPileMove implements SolitaireMove {
    
    /**
     * The index into the stock pile before the move is applied.
     * 
     * <p>For a legal move, this will be equal to the size of the stock pile.</p>
     */
    private final int sourceIndex;
    
    /**
     * Creates a new move that recycles the stock pile.
     * 
     * @param sourceIndex the index into the stock pile before the move
     */
    public RecycleStockPileMove(final int sourceIndex) {
        super();
        this.sourceIndex = sourceIndex;
    }
    
    /**
     * Returns the index into the stock pile before the move is applied.
     * 
     * @return the stock pile index beforehand
     */
    public int getSourceIndex() {
        return sourceIndex;
    }
    
    @Override
    public boolean hasCards() {
        return false;
    }
    
    @Override
    public List<Card> getCards() {
        throw new UnsupportedOperationException();
    }
    
    @Override
    public Board apply(final Board board) {
        return new Board(board.columns, board.stockPile, 0, board.foundation);
    }
    
    @Override
    public int hashCode() {
        return 0x5f23bc91 ^ sourceIndex;
    }
    
    @Override
    public boolean equals(final Object obj) {
        if (obj instanceof RecycleStockPileMove) {
            final RecycleStockPileMove move = (RecycleStockPileMove) obj;
            return sourceIndex == move.sourceIndex;
        } else {
            return false;
        }
    }
    
    @Override
    public String toString() {
        return "Recycle stock pile from index " + sourceIndex + " to the beginning.";
    }
    
}
